pkg/util/render.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"bytes"
+	"fmt"
+	"path/filepath"
+	"strings"
+	"text/template"
+)
+
+const (
+	templateName = "mainTemplate"
+)
+
+// Renderer is golang template renders
+type Renderer struct{}
+
+// NewRenderer returns a new renderer implementation
+func NewRenderer() *Renderer {
+	return &Renderer{}
+}
+
+// Render is responsible for actually rendering the template
+func (r *Renderer) Render(content string, context map[string]interface{}, snippets map[string]string) (rendered string, err error) {
+	// @step: create the template
+	tm := template.New(templateName)
+	if _, err = tm.Funcs(r.templateFuncsMap(tm)).Parse(content); err != nil {
+		return
+	}
+	tm.Option("missingkey=error")
+
+	// @step: add the snippits into the mix
+	for filename, snippet := range snippets {
+		name := filepath.Base(filename)
+		if name == templateName {
+			return "", fmt.Errorf("snippet cannot have the same name as the template: %s", name)
+		}
+		if _, err = tm.New(name).Parse(snippet); err != nil {
+			return rendered, fmt.Errorf("unable to parse snippet: %s, error: %s", filename, err)
+		}
+	}
+
+	// @step: render the actual template
+	writer := new(bytes.Buffer)
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("failed to parse template, error: %s", r)
+		}
+	}()
+	if err = tm.ExecuteTemplate(writer, templateName, context); err != nil {
+		return
+	}
+
+	return writer.String(), nil
+}
+
+// templateFuncsMap returns a map if the template functions for this template
+func (r *Renderer) templateFuncsMap(tm *template.Template) template.FuncMap {
```

Comment:
Wonder if we should use a library for this. What is helm using?

Reply:
@chrislovecnm  Helm is based on `text/template`  to be able to use gotemplate i guess. The helm templating code is going a bit more then just what the `text/template` package gives and were not that portable. i guess with some effort the helm templating can be a package but that is not without effort.